### PR TITLE
Additional work on issue 19453

### LINKED ISSUE #19453: Remove UserProxy from Hibernate
We need to remove the legacy code used in the UserProxy from hibernate and all the classes that use it.

### PR DESCRIPTION
**Pull request by nollymar**
These changes include:

* Refactoring to get rid of hibernate from `user_`
* User additional information is persisted as a json field in the `user_` table instead of  `user_proxy`
* `user_proxy` table was deleted and the data on this table was migrated to `user_` and stored as JSON

### CODE REVIEW

**Comment by dsilvam** [comment]
I recommend running the tests against all DBs

**Review by dsilvam — APPROVED** [review]

**Review by erickgonzalez — APPROVED** [review]

**Comment by dsilvam** [comment]
@jdotcms @wezell according to @nollymar this one is waiting for your review

File: dotCMS/src/main/java/com/dotcms/util/user/LiferayUserTransformer.java [hunk 1]
```diff
@@ -153,7 +156,11 @@ public User fromMap(final Map<String, Object> map) {
             if (null != map.get("delete_date")) {
                 user.setDeleteDate((java.util.Date) map.get("delete_date"));
             }
-
+            if (null != map.get("additional_info")) {
+                user.setAdditionalInfo(Try.of(() -> new ObjectMapper()
+                        .readValue((String) map.get("additional_info"), HashMap.class))
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/util/user/LiferayUserTransformer.java:161]: Not sure if we should create a new Object Mapper every time
We have experimented issues with the same strategy on import and PP with XStream

Take a look to DotObjectMapperProvider may be

File: dotCMS/src/main/java/com/dotmarketing/business/FactoryLocator.java [hunk 2]
```diff
@@ -120,6 +120,15 @@ public static UserFactory getUserFactory(){
     	return (UserFactory)getInstance(FactoryIndex.USER_FACTORY);
     }
 
+    /**
+     * @deprecated Use {@link FactoryLocator#getUserFactory()} instead
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/FactoryLocator.java:124]: good

File: dotCMS/src/main/java/com/dotmarketing/business/UserFactoryImpl.java [hunk 3]
```diff
@@ -0,0 +1,726 @@
...
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.logging.log4j.util.Strings;
+
+public class UserFactoryImpl implements UserFactory {
+
+    private UserCache uc;
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/UserFactoryImpl.java:32]: rename to userCache

Comment (jdotcms) [reply dotCMS/src/main/java/com/dotmarketing/business/UserFactoryImpl.java:32]: set to final

File: dotCMS/src/main/java/com/dotmarketing/business/UserFactoryImpl.java [hunk 4]
```diff
@@ -0,0 +1,726 @@
...
+    public UserFactoryImpl() {
+        uc = CacheLocator.getUserCache();
+    }
+
+    @Override
+    public User loadDefaultUser(final Company company) throws DotDataException, NoSuchUserException {
+        try {
+            return loadUserById(company.getCompanyId() + "." + User.DEFAULT);
+        } catch (NoSuchUserException e) {
+            Logger.debug(this, "Default user not found attempting to create by updating company");
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/UserFactoryImpl.java:49]: Encapsulate this in a method

File: dotCMS/src/main/java/com/dotmarketing/business/UserFactoryImpl.java [hunk 5]
```diff
@@ -0,0 +1,726 @@
...
+            dotConnect.setMaxRows(limit);
+        }
+        final List<Map<String, Object>> results = dotConnect.loadResults();
+
+        // Since limit is a small number, convert each row to appropriate entity
+        final List<User> users = new ArrayList<User>();
+
+        for (final Map<String, Object> hash : results) {
+            final String userId = (String) hash.get("userid");
+            final User u = loadUserById(userId);
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/UserFactoryImpl.java:239]: rename to user

File: dotCMS/src/main/java/com/dotmarketing/business/UserFactoryImpl.java [hunk 6]
```diff
@@ -0,0 +1,726 @@
...
+                .addParam(user.getLoginIP()).addParam(user.getLastLoginDate()).addParam(user.getLastLoginIP())
+                .addParam(user.getFailedLoginAttempts()).addParam(user.getAgreedToTermsOfUse())
+                .addParam(user.getActive()).addParam(user.getDeleteInProgress()).addParam(user.getDeleteDate())
+                .addParam(user.getAdditionalInfo()!=null ? mapper.writeValueAsString(user.getAdditionalInfo()): "{}");
+    }
+
+    @Override
+    public boolean userExistsWithEmail(final String email, final String companyId) throws DotDataException {
+
+        User u;
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/UserFactoryImpl.java:373]: rename to user

File: dotCMS/src/main/java/com/dotmarketing/business/UserHelper.java [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+import com.liferay.util.InstancePool;
+import com.liferay.util.Time;
+import com.liferay.util.Validator;
+import java.util.Date;
+import java.util.Locale;
+import org.apache.commons.lang3.StringUtils;
+
+public class UserHelper {
+
+    private UserAPI userAPI;
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/UserHelper.java:42]: set to final

File: dotCMS/src/main/java/com/dotmarketing/business/UserHelper.java [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+    public User getUserObject(final String companyId, final boolean autoUserId,
+            final String userId, final boolean autoPassword,
+            final String password1, final String password2,
+            final boolean passwordReset, final String firstName,
+            final String middleName, final String lastName,
+            final String nickName, final boolean male, final Date birthday,
+            final String emailAddress, final Locale locale)
+            throws DotDataException, SystemException, PortalException {
+
+        boolean alwaysAutoUserId = GetterUtil.getBoolean(
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/business/UserHelper.java:65]: Set to final

**Review by jdotcms — APPROVED** [review]

**Review by wezell — CHANGES_REQUESTED** [review]
Nolly, the changes are headed in the right direction.  I added some questions and requested a few changes in the code.

File: dotCMS/src/main/java/com/dotcms/util/user/LiferayUserTransformer.java (hunk 1 above)
Comment (nollymar) [reply dotCMS/src/main/java/com/dotcms/util/user/LiferayUserTransformer.java:161]: You're right. Good catch!

Comment (nollymar) [reply dotCMS/src/main/java/com/dotcms/util/user/LiferayUserTransformer.java:162]: Done

File: dotCMS/src/main/java/com/dotmarketing/business/UserFactoryImpl.java (hunk 3 above)
Comment (nollymar) [reply dotCMS/src/main/java/com/dotmarketing/business/UserFactoryImpl.java]: Done

File: dotCMS/src/main/java/com/dotmarketing/business/UserFactoryImpl.java (hunk 4 above)
Comment (nollymar) [reply dotCMS/src/main/java/com/dotmarketing/business/UserFactoryImpl.java:49]: Done

File: dotCMS/src/main/java/com/dotmarketing/business/UserFactoryImpl.java (hunk 5 above)
Comment (nollymar) [reply dotCMS/src/main/java/com/dotmarketing/business/UserFactoryImpl.java]: Done

File: dotCMS/src/main/java/com/dotmarketing/business/UserFactoryImpl.java (hunk 6 above)
Comment (nollymar) [reply dotCMS/src/main/java/com/dotmarketing/business/UserFactoryImpl.java]: Done

File: dotCMS/src/main/java/com/dotmarketing/business/UserHelper.java (hunk 7 above)
Comment (nollymar) [reply dotCMS/src/main/java/com/dotmarketing/business/UserHelper.java]: Done

File: dotCMS/src/main/java/com/dotmarketing/business/UserHelper.java (hunk 8 above)
Comment (nollymar) [reply dotCMS/src/main/java/com/dotmarketing/business/UserHelper.java]: Done

File: dotCMS/src/main/java/com/dotcms/business/LazyUserAPIWrapper.java [hunk 9]
```diff
@@ -135,7 +135,8 @@ public long getCountUsersByNameOrEmailOrUserID(String filter, boolean includeAno
     }
 
     @Override
-    public Optional<String> getUserIdByIcqId(final String icqId) throws DotInvalidTokenException {
+    public Optional<String> getUserIdByIcqId(final String icqId)
```
[the remark this answers is not in the document]
Comment (nollymar) [reply dotCMS/src/main/java/com/dotcms/business/LazyUserAPIWrapper.java]: Done

File: dotCMS/src/main/java/com/dotmarketing/business/UserAPI.java [hunk 10]
```diff
@@ -440,6 +440,7 @@ public long getCountUsersByNameOrEmailOrUserID(String filter, boolean includeAno
 	 * @param icqId icqId to search for
 	 * @return userId that the icqId is associated
 	 */
-	public Optional<String> getUserIdByIcqId(final String icqId) throws DotInvalidTokenException;
+	public Optional<String> getUserIdByIcqId(final String icqId)
```
[the remark this answers is not in the document]
Comment (nollymar) [reply dotCMS/src/main/java/com/dotmarketing/business/UserAPI.java]: Done

File: dotCMS/src/main/java/com/dotmarketing/business/UserAPIImpl.java [hunk 11]
```diff
@@ -140,38 +144,71 @@ public String encryptUserId(final String userId) throws DotStateException{
...
+        return getCountUsersByName(filter, null);
     }
 
     @CloseDBIfOpened
     @Override
-    public long getCountUsersByName(String filter, List<Role> roles) throws DotDataException {
-        return userFactory.getCountUsersByName(filter, roles);
+    public long getCountUsersByName(String filter, List<Role> roles) {
+        return userFactory.getCountUsersByName(filter, roles,
+                APILocator.getCompanyAPI().getDefaultCompany().getCompanyId());
```
[the remark this answers is not in the document]
Comment (nollymar) [reply dotCMS/src/main/java/com/dotmarketing/business/UserAPIImpl.java]: Done

File: dotCMS/src/main/java/com/dotmarketing/business/UserAPIImpl.java [hunk 12]
```diff
@@ -140,38 +144,71 @@ public String encryptUserId(final String userId) throws DotStateException{
...
+        user.setNew(false);
+        return userFactory.save(user);
     }
 
     @CloseDBIfOpened
     @Override
     public User getDefaultUser() throws DotDataException {
         try {
-            return userFactory.loadDefaultUser();
+            return userFactory.loadDefaultUser(APILocator.getCompanyAPI().getDefaultCompany());
```
[the remark this answers is not in the document]
Comment (nollymar) [reply dotCMS/src/main/java/com/dotmarketing/business/UserAPIImpl.java:208]: I removed the company filter from everywhere, except for the `loadDefaultUser`, as it is used to populate the default user object

File: dotCMS/src/main/java/com/dotmarketing/business/UserAPIImpl.java [hunk 13]
```diff
@@ -574,13 +612,13 @@ public void saveAddress(User user, Address ad, User currentUser, boolean respect
...
         }
-        userFactory.saveAddress(user, ad);
+        userFactoryLiferay.saveAddress(user, ad);
     }
 
     @CloseDBIfOpened
     @Override
     public Address loadAddressById(String addressId, User currentUser, boolean respectFrontEndRoles) throws DotDataException, DotSecurityException {
-        Address ad = userFactory.loadAddressById(addressId);
+        Address ad = userFactoryLiferay.loadAddressById(addressId);
```
[the remark this answers is not in the document]
Comment (nollymar) [reply dotCMS/src/main/java/com/dotmarketing/business/UserAPIImpl.java:618]: Removed

File: dotCMS/src/main/java/com/dotmarketing/business/UserAPIImpl.java [hunk 14]
```diff
@@ -593,7 +631,7 @@ public void deleteAddress(Address ad, User currentUser, boolean respectFrontEndR
         if(!permissionAPI.doesUserHavePermission(userProxyAPI.getUserProxy(ad.getUserId(),APILocator.getUserAPI().getSystemUser(), false), PermissionAPI.PERMISSION_EDIT, currentUser, respectFrontEndRoles)){
             throw new DotSecurityException("User doesn't have permission to userToDelete the user which is trying to be saved");
         }
-        userFactory.deleteAddress(ad);
+        userFactoryLiferay.deleteAddress(ad);
```
[the remark this answers is not in the document]
Comment (nollymar) [reply dotCMS/src/main/java/com/dotmarketing/business/UserAPIImpl.java]: Removed

File: dotCMS/src/main/java/com/dotmarketing/business/UserFactory.java [hunk 15]
```diff
@@ -1,350 +1,222 @@
...
+
     /**
      * Will create user if it doesn't exist
-     * @return
-     * @throws DotDataException
-     * @throws NoSuchUserException
      */
-    protected abstract User loadDefaultUser() throws DotDataException, NoSuchUserException;
-    
+    User loadDefaultUser(final Company company) throws DotDataException, NoSuchUserException;
```
[the remark this answers is not in the document]
Comment (nollymar) [reply dotCMS/src/main/java/com/dotmarketing/business/UserFactory.java:46]: I removed the company filter from everywhere, except from the `loadDefaultUser`, as it is used to populate the default user object

**Pull request merged**